src/mlpack/tests/main_tests/krann_test.cpp

```diff
@@ -0,0 +1,335 @@
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 5);
+  SetInputParam("tau", (double) -1); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Make sure that dimensions of the neighbors and distances matrices are correct
+ * given a value of k.
+ */
+BOOST_AUTO_TEST_CASE(KRANNOutputDimensionTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of  top tau %ile reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 5);
+
+  mlpackMain();
+
+  // Check the neighbors matrix has 5 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>
+      ("neighbors").n_rows, 5);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>
+      ("neighbors").n_cols, 100);
+
+  // Check the distances matrix has 10 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_rows, 5);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_cols, 100);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(KRANNModelReuseTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  arma::mat queryData;
+  queryData.randu(3, 90); // 90 points in 3 dimensions.
+
+  // Random input, some k <= number of  top tau %ile reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("query", queryData);
+  SetInputParam("k", (int) 5);
+
+  mlpackMain();
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  RANNModel* output_model;
+  neighbors = std::move(CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  distances = std::move(CLI::GetParam<arma::mat>("distances"));
+  output_model = std::move(CLI::GetParam<RANNModel*>("output_model"));
+
+  // Reset passed parameters.
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+
+  // Input saved model, pass the same query and keep k unchanged.
+  SetInputParam("input_model", output_model);
+  SetInputParam("query", queryData);
+
+  mlpackMain();
+
+  // Check that initial output matrices and the output matrices using
+  // saved model are equal.
+  CheckMatrices(neighbors, CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  CheckMatrices(distances, CLI::GetParam<arma::mat>("distances"));
+}
+
+/**
+  * Ensure that different leaf sizes give different results.
+ */
+BOOST_AUTO_TEST_CASE(KRANNDifferentLeafSizes)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of  top tau %ile reference points.
+  SetInputParam("reference", referenceData);
+  SetInputParam("k", (int) 5);
+  SetInputParam("leaf_size", (int) 1);
```

Comment:
Rank-approximate nearest neighbor search has a random component, so if you run this test you should also specify the random seed.  Take a look at `main_tests/cf_test.cpp` to see how you can use `mlpack::math::FixedRandomSeed()` to specify a fixed random seed. :+1: